File: backend/app/limiters/token_bucket.py
from __future__ import annotations

import math

from redis.asyncio import Redis

from app.limiters.base import Decision


class TokenBucketLimiter:
    algorithm = "token_bucket"

    async def allow(
        self,
        redis_client: Redis,
        policy_id: str,
        client_id: str,
        now_ms: int,
        params: dict[str, object],
    ) -> Decision:
        capacity = float(params["capacity"])
        refill_rate_per_sec = float(params["refill_rate_per_sec"])
        tokens_per_request = float(params.get("tokens_per_request", 1))

        key = f"rl:token:{policy_id}:{client_id}"
        tokens_raw, last_refill_raw = await redis_client.hmget(key, "tokens", "last_refill_ms")

        if tokens_raw is None or last_refill_raw is None:
            tokens = capacity
            last_refill_ms = now_ms
        else:
            tokens = float(tokens_raw)
            last_refill_ms = int(float(last_refill_raw))

        elapsed_sec = max((now_ms - last_refill_ms) / 1000.0, 0.0)
        refilled_tokens = min(capacity, tokens + (elapsed_sec * refill_rate_per_sec))

        allowed = refilled_tokens >= tokens_per_request
        if allowed:
            remaining_tokens = refilled_tokens - tokens_per_request
            reason = None
            retry_after_ms = None
        else:
            remaining_tokens = refilled_tokens
            reason = "rate_limit_exceeded"
            needed_tokens = tokens_per_request - refilled_tokens
            retry_after_ms = max(int(math.ceil((needed_tokens / refill_rate_per_sec) * 1000)), 1)

        ttl_sec = max(int(math.ceil(capacity / refill_rate_per_sec)) + 2, 2)
        await redis_client.hset(
            key,
            mapping={
                "tokens": f"{remaining_tokens:.8f}",
                "last_refill_ms": now_ms,
            },
        )
        await redis_client.expire(key, ttl_sec)

        remaining_requests = int(max(math.floor(remaining_tokens / tokens_per_request), 0))
        return Decision(
            allowed=allowed,
            reason=reason,
            remaining=remaining_requests,
            retry_after_ms=retry_after_ms,
            algorithm_state={"tokens": round(remaining_tokens, 4)},
        )

File: gateway/internal/limiter/token_bucket.go
package limiter

import (
	"context"
	"fmt"
	"math"
)

type TokenBucketLimiter struct {
	store RuntimeCounter
}

func NewTokenBucketLimiter(store RuntimeCounter) *TokenBucketLimiter {
	return &TokenBucketLimiter{store: store}
}

func (l *TokenBucketLimiter) Allow(ctx context.Context, policyID, clientID string, nowMS int64, params map[string]any) (Decision, error) {
	capacity, err := readPositiveInt(params, "capacity")
	if err != nil {
		return Decision{}, err
	}
	refillRate, err := readPositiveFloat(params, "refill_rate_per_sec")
	if err != nil {
		return Decision{}, err
	}
	tokensPerRequest := 1
	if value, ok := params["tokens_per_request"]; ok {
		tokensPerRequest, err = readPositiveInt(map[string]any{"tokens_per_request": value}, "tokens_per_request")
		if err != nil {
			return Decision{}, err
		}
	}

	key := fmt.Sprintf("rl:token_bucket:%s:%s", policyID, clientID)
	ttlSec := int(math.Ceil(float64(capacity)/refillRate)) + 1
	if ttlSec < 2 {
		ttlSec = 2
	}

	result, err := l.store.EvalTokenBucket(ctx, key, nowMS, capacity, refillRate, tokensPerRequest, ttlSec)
	if err != nil {
		return Decision{}, err
	}

	remainingBudget := int(math.Floor(result.Tokens / float64(tokensPerRequest)))
	if remainingBudget < 0 {
		remainingBudget = 0
	}

	state := map[string]any{
		"count":                remainingBudget,
		"window_start_ms":      nowMS,
		"window_size_sec":      0,
		"state_schema_version": 1,
		"tokens":               result.Tokens,
		"capacity":             capacity,
		"refill_rate_per_sec":  refillRate,
		"tokens_per_request":   tokensPerRequest,
		"last_refill_ms":       result.LastRefillMS,
	}

	if result.Allowed {
		return Decision{
			Allowed:        true,
			Remaining:      &remainingBudget,
			AlgorithmState: state,
		}, nil
	}

	reason := "rate_limit_exceeded"
	zero := 0
	retryAfter := result.RetryAfterMS
	return Decision{
		Allowed:        false,
		Reason:         &reason,
		Remaining:      &zero,
		RetryAfterMS:   &retryAfter,
		AlgorithmState: state,
	}, nil
}

func readPositiveFloat(input map[string]any, key string) (float64, error) {
	value, ok := input[key]
	if !ok {
		return 0, fmt.Errorf("missing required field %s", key)
	}
	switch v := value.(type) {
	case float64:
		if v <= 0 {
			return 0, fmt.Errorf("field %s must be > 0", key)
		}
		return v, nil
	case int:
		if v <= 0 {
			return 0, fmt.Errorf("field %s must be > 0", key)
		}
		return float64(v), nil
	default:
		return 0, fmt.Errorf("field %s must be numeric", key)
	}
}
